### ISSUE TITLE
Create no-shadow pure function arguments rule

### ISSUE THREAD

**Issue by mike-robertson** [eslint/eslint#6307]
**When does this rule warn? Please describe and show example code:**
This rule would allow you to use the same variable names if they are calling the code when you want a pure function which guarantees no side effects (for ease of reading and logic).

ex:
```js
let sortedCount = 0;
sortedCount = myPureIncrement(sortedCount); //1
sortedCount = myPureIncrement(sortedCount); //2

const myPureIncrement = (sortedCount) => {
  return sortedCount + 1;
};
```
```js
let sortedCount = 0;
sortedCount = myPureIncrement(sortedCount); //1
sortedCount = myPureIncrement(sortedCount); //2

const myPureIncrement = (c) => { 
// I have to come up with a new name for all these functions
// where I don't even want to have a chance of mutating the item in the function
  return c + 1;
};
```
**Is this rule preventing an error or is it stylistic?**
Allows you to write pure functions which use the same variable name as the argument being passed to it. It's annoying to have to come up with all sorts of new names when factoring out logic into functions while wanting to maintain the pureness of the function.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**
I don't know. I just want to have normal shadowing unless the variable I call the function with is the same name as the argument the function is creating to store that value.

**Are you willing to create the rule yourself?**
I'm not sure.

**Comment by ilyavolodin** [comment]
I don't think this is really doable with static code analysis. Determining if the function is pure would require a huge amount of code and I don't think it's worth it (it also can't be done with 100% certainty). And even in pure functions it's not a good idea to shadow arguments. You can always turn the rule off completely or just around your pure function with comments.

**Comment by Mikejoro** [comment]
I don't think the rule would be to detect the pureness of the Function, but rather detect whether the Function is called with an argument of the same name as the parameter which receives it. 

There's no downside to shadowing the variable then because I don't want the ability to mutate it in the parent scope, and this prevents the mutation and forces you to use a return value with an assigned to mutate the scope. Sort of like preferring x = x + 1 vs. x++

Maybe it's just a weird style of how I code and wouldn't be useful to others though.

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
